quota/quota.go

```diff
@@ -0,0 +1,315 @@
+// from the snapd friendly group name, mainly in the case that the group is a
+// sub group.
+func (grp *Group) SliceFileName() string {
+	if grp.ParentGroup == "" {
+		// root group name, then the slice unit is just "<name>.slice"
+		return fmt.Sprintf("%s.slice", grp.Name)
+	}
+
+	// otherwise we need to track back to get all of the parent elements
+	grpNames := []string{}
+	parentGrp := grp.parentGroup
+	for parentGrp != nil {
+		grpNames = append([]string{parentGrp.Name}, grpNames...)
+		parentGrp = parentGrp.parentGroup
+	}
+
+	buf := &bytes.Buffer{}
+	for _, parentGrpName := range grpNames {
+		fmt.Fprintf(buf, "%s-", parentGrpName)
+	}
+	fmt.Fprintf(buf, "%s.slice", grp.Name)
+	return buf.String()
+}
+
+// allow only alphanumerics at least one character long, and not starting with
+// a number
+var validGroupNameRegexp = regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9]*$`)
+
+func (grp *Group) validate() error {
+	// check that the name is a simple alphanumeric name
+	if !validGroupNameRegexp.MatchString(grp.Name) {
+		return fmt.Errorf("group name %q contains invalid characters (valid names are alphanumeric starting with a letter)", grp.Name)
+	}
+
+	// check if the name is reserved for future usage
+	switch grp.Name {
+	case "root", "system", "snapd", "user":
+		return fmt.Errorf("group name %q reserved", grp.Name)
+	}
+
+	if grp.MemoryLimit == 0 {
+		return fmt.Errorf("group memory limit must be non-zero")
+	}
+
+	// TODO: probably there is a minimum amount of bytes here that is
+	// technically usable/enforcable, should we check that too?
+
+	// check that if this is a sub-group, then the parent group has enough space
+	// to accommodate this new group (we assume that other existing sub-groups
+	// in the parent group have already been validated)
+	if grp.parentGroup != nil {
+		totalAvailable := quantity.Size(0)
+		for _, child := range grp.parentGroup.subGroups {
+			totalAvailable += child.MemoryLimit
+		}
+		if totalAvailable+grp.MemoryLimit > grp.parentGroup.MemoryLimit {
+			remaining := grp.parentGroup.MemoryLimit - totalAvailable
+			return fmt.Errorf("sub-group memory limit of %s is too large to fit inside remaining quota space %s for parent group %s", grp.MemoryLimit.IECString(), remaining.IECString(), grp.parentGroup.Name)
+		}
+	}
+
+	return nil
+}
+
+// NewSubGroup creates a new sub group under the current group
+func (grp *Group) NewSubGroup(name string, memLimit quantity.Size) (*Group, error) {
+	subGrp := &Group{
+		Name:        name,
+		MemoryLimit: memLimit,
+		ParentGroup: grp.Name,
+		parentGroup: grp,
+	}
+
+	if err := subGrp.validate(); err != nil {
+		return nil, err
+	}
+
+	// also double check that the sub group name is not the same as that of the
+	// parent, this is fine in systemd world, but in snapd we want unique quota
+	// groups
+	if name == grp.Name {
+		return nil, fmt.Errorf("cannot use same name %q for sub group as parent group", name)
```

Comment:
That's ok but I guess we need more checks - where are we going to ensure the name is globally unique for all the groups?

Reply:
The intention is that there will be checks inside servicestate where we store quotas inside state.json as a map of the name to the group object, such that it will be immediately obvious whether a new user-provided name already exists as another group or not